proposals/named-tuples.md

```diff
@@ -0,0 +1,112 @@
+the tuple type becomes central to the design of a piece of your program, the more the structural
+typing becomes a problem and the desire for traditional C# nominal typing becomes prominent.
+
+This is generally the point where most users will want to evolve their tuple type into a proper
+named type. Unfortunately, this is extremely laborious. Now the following must be defined manually:
+
+  1. `ItemX` properties for each positional member of the type
+  2. Optionally, custom named properties for each positional member of the type
+  3. A constructor to assign each of the input parameters
+  4. A deconstructor to support pattern matching
+  4. Memberwise equality, including `GetHashCode` and `IEquatable`
+
+While the associated `data` named types makes some of this easier, the positionality and tuple
+characteristics are not addressed and are orthogonal to the former proposal.
+
+# Proposal
+
+To solve these problems, I propose what I'm calling "named tuples."
+
+A named tuple has the following syntax for classes, with an analogous definition for structs:
+
+```antlr
+class_declaration
+    : attributes? class_modifier* 'partial'? 'class' identifier type_parameter_list?
+      parameter_list? class_base? type_parameter_constraints_clause* (class_body | ';')?
+    ;
+
+parameter_list
+    : '(' parameters ')'
+    ;
+
+parameters
+    : attributes? parameter
+    | parameters ',' attributes? parameter
+    ;
+
+parameter
+    : 'readonly'? identifier identifier?
+    ; 
+```
+
+For example, a Cartesian "point" class could like the following as a named tuple:
+
+```C#
+class Point(int, int);
+
+void M()
+{
+    var p = new Point(0, 0);
+}
+```
+
+Like tuples, names of the elements are optional. If no names are provided, a consumer
+can refer to automatically generated `ItemX` properties, as well as use the automatically
+generated Deconstruct method, e.g.
+
+```C#
+void M()
+{
+    var (x, y) = new Point(0, 0);
+}
+```
+
+In the simplest form, the body of the class is also optional.
+
+## Generated members
+
+From the definition of `data` named types, you can see that tuples already conform to many
+of the same semantics. The same is true of named tuples. Equality is generated like
+for `data` named types, but the `data` keyword is not needed. One difference between equality
+in named tuples and equality in anonymous tuples is that anonymous tuples are structurally
+equal, while named tuples follow the C# standard of nominal equality for named types. This
+means that one named tuple instance can only be equal to another named tuple instance if they
+are the same type, not if they just have equal members.
+
+In addition to the equality members, named tuples also generate
+
+  1. A constructor corresponding to the parameter list in the type definition.
+  2. `ItemX` properties for each member
+  3. If a parameter is named in the parameter list, a public property with the same name
+     that gets/sets that member.
```

Comment:
in the past, there's been contention about naming properties the same name as parameters.

Reply:
I've decided to not care.